### PR TITLE
updating the checkpointer and checkpoint loader

### PR DESCRIPTION
**Pull request by AliGhadirzadeh**
Made some updates to the Checkpointer and Checkpointloader as follows:

* Removed the extra input arguments for loading networks and optimizers. Now, you can only load a list of callbacks. However, added an additional input to the Checkpointloader class. When set to True, this input instructs the loader to exclusively load the neural network parameters, disregarding the state_dict of other callbacks. This feature is useful in scenarios where you need to copy a network from one callback to another or when you prefer not to load other callback parameters.

Both Checkpointer and Checkpointloader rely on a list of callbacks rather than a dictionary. To ensure the correct loading of callbacks, we store a corresponding list of names when saving the checkpoints. These names serve as references during the loading process. The names are loaded from the callback names. Callbacks without a name field are discarded. An error will be raised in case repetitive names are given.

### CODE REVIEW

File: emote/callback.py [hunk 1]
```diff
@@ -211,7 +211,7 @@ def end_training(self, *args, **kwargs):
     def state_dict(self) -> Dict[str, Any]:
         return {}
 
-    def load_state_dict(self, state_dict: Dict[str, Any]):
+    def load_state_dict(self, state_dict: Dict[str, Any], *args, **kwargs):
```
Comment (tgolsson) [inline emote/callback.py:214]: I don't think this has to be generic! We have previously used three different flags and I'd prefer to maintain that flow:

* load_weights
* load_optimizers
* load_hparams - this is a bit badly named, but essentially inf-steps and other things like that that aren't in either of the above categories! 

I think that would make more sense than having flags that are "only x". What if I want only optimizers? Only schedules? Weights and optimizers but no schedules? :-)

**Review by tgolsson — CHANGES_REQUESTED** [review]
A minor change request to ensure we don't build ourselves into a corner! I think exclusive flags get quite messy over time, so think clear opt ins/outs on individual parts make more sense. I think that will make some other code cleaner too. :-)

**Review by singhblom — APPROVED** [review]
This looks clean and good to me.

File: emote/callbacks/loss.py [hunk 2]
```diff
@@ -52,12 +52,21 @@ def state_dict(self):
...
+    ):
+        if self.network and load_weights:
             self.network.load_state_dict(state_dict.pop("network_state_dict"))
-        super().load_state_dict(state_dict)
+
+        if self.optimizer and load_optimizers:
+            self.optimizer.load_state_dict(state_dict.pop("optimizer_state_dict"))
+
+        if load_hparams:
+            super().load_state_dict(state_dict)
```
Comment (tgolsson) [inline emote/callbacks/loss.py:69]:
```suggestion
        super().load_state_dict(state_dict, load_weights, load_optimizers, load_hparams)
```

We should always call this function, and always pass the flags. :-)

Comment (AliGhadirzadeh) [reply emote/callbacks/loss.py]: Thanks for it! it is fixed now.

**Review by tgolsson — APPROVED** [review]
Awesome! LGTM.
